### PR TITLE
[Feature] Add import and export of anonymised data tables

### LINKED ISSUE #2: Add the ability to generate anonymised exports
### Background

The current process for creating our database dumps strips out sensitive tables which is great for development environments, due to it reducing the size of the dump itself. For diagnosing and fixing some client issues, however, the volume of data is key. Any reasonable security policy will prevent developers from importing sensitive customer data into development environments, so we need to support obfuscation of certain columns in specific tables.

### Related Projects

* [divanteltd/anonymizer](https://github.com/divanteltd/anonymizer) (ruby)
* [mavenlink/my_obfuscate](https://github.com/mavenlink/my_obfuscate) (ruby)

### Requirements

* As a developer I should be able to specify which columns in tables that I want anonymised
* As a developer I should be able to keep anonymisation configuration in my project's version control system
* As a developer I should be able to specify whether or not I want to generate a stripped database backup or an anonymised backup

### Approach
The plan to start off with is to use one of the tools in the _Related Projects_ section to do the anonymisation for us. Stream reading and modifying an SQL dump is a project in itself, so it would be wise for us to stay focused on the overall role of this project and tool. This does introduce a dependency on having ruby available on the machine (unless we can find a PHP equivalent), but that’s ok for the time being. Medium to long term, narrowing down the requirements to just PHP is the aim.

### Tasks

- [x] Generate an anonymised export
- [x] Import an anonymised export
- [x] Upload exports to S3
- [x] Download exports from S3
- [x] Allow anonymization rules to be driven by configuration
- [x] Allow for the specification of a project specific configuration file

### PR DESCRIPTION
**Pull request by punkstar**
- [Feature] Add ExportCommand to support the generation of the sensitive data
- [Feature] Add ImportCommand to support the import of the sensitive data
- [Feature] Add Formatter implementations to perform specific operations on certain row values
- [Feature] Add anonymizer/tables section to configuration where formatter rules are defined
- [Enhancement] Add a DI container (php-di)
- [Enhancement] Refactored configuration to support merging and multiple sources (multiple files and the Symfony\Console InputInterface)
- [Enhancement] Centralised constants for global configuration options
- [Enhancement] Improved ConfigureCommand to output the contents of your configuration file before asking you for new values
- [Enhancement] Use a file size unit (b, kb, mb, gb) that's applicable to the size of the file when using the LsCommand
- [Enhancement] Add Shell\* wrappers for shell commands to avoid direct execs and process management in *Command classes

Closes #2.

### CODE REVIEW

File: src/Service/DatabaseFactory.php [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+     * @return DatabaseInterface
+     * @throws \Psr\Container\ContainerExceptionInterface
+     * @throws \Psr\Container\NotFoundExceptionInterface
+     */
+    public static function create(Container $container)
+    {
+        switch ($container->get('database_adapter')) {
+            case 'shell':
+            default:
+                $instance = $container->make(\Meanbee\Magedbm2\Service\Database\Shell::class);
```
Comment (tgerulaitis) [inline src/Service/DatabaseFactory.php]: It would be nice to eventually move all of these to `di.php` directly.

File: src/Service/Storage/S3.php [hunk 2]
```diff
@@ -317,8 +334,15 @@ protected function addInputOptions(Application $app)
         $definition->addOption(new InputOption(
             "bucket",
             null,
-            InputOption::VALUE_REQUIRED,
-            "S3 bucket"
+            InputOption::VALUE_OPTIONAL,
```
Comment (tgerulaitis) [inline src/Service/Storage/S3.php]: Do we want people to be able to specify the `--bucket` option without giving it a value? Or am I misunderstanding how `VALUE_OPTIONAL` works?

File: phpmd.xml [hunk 3]
```diff
@@ -10,8 +10,14 @@
...
         <exclude name="ElseExpression"/>
+        <exclude name="StaticAccess"/>
+    </rule>
+    <rule ref="rulesets/codesize.xml">
+        <exclude name="CyclomaticComplexity" />
+    </rule>
+    <rule ref="rulesets/design.xml">
+        <exclude name="CouplingBetweenObjects" />
+
     </rule>
```
[the remark this answers is not in the document]
Comment (punkstar) [reply phpmd.xml]: I was finding it difficult to meeting the requirements of the linting in a large number of places (which is probably a reflection on the quality of how I've implemented this, to be honest), so just decided to disable globally.

I don't remember how extensive it was. I'll reintroduce them and see what it's looking like.

File: src/Anonymizer/Formatter/Address/RegionId.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+class RegionId implements FormatterInterface
+{
+    /**
+     * @param $value
+     * @param array $rowContext
+     * @return string
+     */
+    public function format($value, array $rowContext)
+    {
+        return 1;
```
[the remark this answers is not in the document]
Comment (punkstar) [reply src/Anonymizer/Formatter/Address/RegionId.php]: So, as you can probably tell from the `$rowContext`, my original plan with formatters in general was being able to set specific formatter values based on the values of other columns.  The region should be based on the region id which should be based on the country id. There are ordering issues around this though which I didn't really want to tackle (didn't really want to make the ordering of the formatters in the configuration file significant, nor worry about what happens when you start merging multiple configuration files) so I avoided it.

I'll return a random number though, based on the data that ships with 2.2.

File: src/Service/DatabaseFactory.php (hunk 1 above)
Comment (punkstar) [reply src/Service/DatabaseFactory.php]: When I first wrote this I did have everything in `di.php` but decided to move it out to factories! I thought it was a bit cleaner, though I'm not a fan of passing the container around.

File: src/Service/Storage/S3.php (hunk 2 above)
Comment (punkstar) [reply src/Service/Storage/S3.php]: Good spot. We might be dealing with a data export so we don't care about `--bucket`, we care about `--data-bucket`. The validation for this is in `validateConfiguration()` a little further down.

**Pull request merged**